### ISSUE TITLE
ZDaemon tables

### ISSUE THREAD

**Issue by Aeyesx** [GitExl/WhackEd4#1]
Hello!

This was supposed to be reguest so sorry if I put it into bad place 
(also sorry for bad eng.)

I was curious if ZDaemon tables (config) could be added?
I could try write them by myself but its huge and complicated,
Authors of original tables have propably 10x more practice with it.
ZDaemon supports up to heretic and hexen (hexen partially but soon will be fully supported too). Using DEHSUPP we can utilize these states and actors to create custom new stuff. What is painy is that there is no other way, just to write everything "manually" in text editors. In comfy editor like WhackEd4 it's gonna be much user friendly.

Thank you!

medis0003@gmail.com
Skype: Aeyesx

**Comment by Aeyesx** [comment]
That is true, no need for sounds section of heretic/hexen/raven. Anyway, how it's going ? I think it'll be great for chick modders, + I'll hope doomer's dehsupp could be official someday (after hexen get finished) in zdaemon.wad

**Comment by GitExl** [comment]
I did some preliminary work on it before the holidays, but right now I'm 
busy with other stuff. It'll get done when it gets done :)

**Comment by GitExl** [comment]
I'm working on completing version 1.1 which includes support for most ZDaemon\ZDoom features (but not yet DEHSUPP). A few things aren't yet 100% clear to me:

The ZDaemon binary has strings for a number of action functions in it. Some are the same as their MBF counterparts and were easy enough to add, but the parameters\usage of the following aren't quite known to me:
* FireRailgun
* FireRailgunLeft
* FireRailgunRight
* Gravity
* HideThing
* LowGravity
* MonsterRail
* NoGravity
* PlayWeaponSound (same as PlaySound but for weapons?)
* RailWait
* SetFloorClip (this one and other Set\UnSet actions change thing flags?)
* SetInvulnerable
* SetReflective
* SetReflectiveInvulnerable
* SetShootable
* UnHideThing
* UnSetFloorClip
* UnSetInvulnerable
* UnSetReflective
* UnSetReflectiveInvulnerable
* UnSetShootable

Likewise in the ZDaemon binary the string "BFG Aiming" appears alongside miscellaneous value strings. What does it do?

Lastly, there's strings for a lot of thing flags in the ZDaemon binary that do not appear in it's default DEHSUPP lump. See http://pastebin.com/8LGFn3Vu for a list of these, a number of which I've identified through the ZDoom source code. Unknown ones are on the very bottom of the list. Am I right in assuming that these thing flags are all supported by ZDaemon's Dehacked parser?

**Comment by Aeyesx** [comment]
Few of them are it's Decorate equialents, few comes from hexen / heretic, BFG Aiming is a variable used in dehacked that simply specify if bfg can shoot Up&Down or does not.

I'll edit you're pastebin in few minutes hopefully explaining what it does...
Either i am not certainly sure about few so I have to ask >) But heck thanks for the effort ! 
It is cool

**Comment by Doomer-** [comment]
Hello,

the codepointers you listed can be found on the zdoom wiki: http://zdoom.org/wiki/ZDoom_codepointer_extensions

PlayWeaponSound is a Zdaemon exclusive codepointer which is similar to PlaySound but it is used on weapon frames, not on actor frames. The reason for needing both codepointers is that codepointers used on actor frames don't work on weapon frames and vice versa. This is an engine limitation of Zdoom 1.22 based ports.

8. Implemented an MBF-style codepointer A_PlayWeaponSound. It's
   similar to A_PlaySound, except that:
     a. it applies to weapon rather than actor frames,
     b. it uses the weapon rather than the body channel,
     c. it uses Args1/Args2 rather than Unknown 1/2 for the sound
        number and attenuation flag respectively (Unknown 1/2 are
        used for the animation in weapon frames).

BFG Aiming specifies wether you can or cannot control the vertical angle of the bfg ball when firing. By default, the BFG ball is always fired straight ahead no matter if freelook is enabled (for some oldschool reasons). This setting allows BFG ball to behave the same as other player projectiles when set to 1.

Only MF1 and MF2 flags are supported by Zdaemon's dehacked, as far as I know. The MF3 and MF4 are Hexen and other special flags, they cannot be used by dehacked. There is also a recent addition:

8. Added a "SYNCMOVEMENT" flag to dehacked; it can be used
    for projectiles having idiosynchratic/randomized movement
    that need frequent "driving" from the server to the clients
    (use it with restraint though: it consumes more bandwidth
    than normal projectiles).

CLIENTSIDE, CLIENTSIDEONLY are the same and explained in one of my previous posts, as well as SERVERONLY.

**Comment by Aeyesx** [comment]
+1 Great explanation as always doomer, I wish to be accurate like that someday either >D

**Comment by GitExl** [comment]
That's very helpful, thanks. I'll add the thing flags that are present in ZDaemon's DEHSUPP, as well as the CLIENTSIDE, CLIENTSIDEONLY, SERVERONLY, SYNCMOVEMENT and FRIENDLY flags. Also the ZDoom codepointer extensions abd OlayWeaponSound. Anything else I consider undocumented and untested, so best not to include at this point.

**Comment by Aeyesx** [comment]
Could you left the "undocumented/untested" list please ? o,o I am curious of them, CLIENTSIDE is de-facto just a synonym for a CLIENTSIDEONLY... I guess use the longer variant as it is used either in zandronum/skulltag for compatibility purposes,

**Comment by Aeyesx** [comment]
Update: 
http://www.zdaemon.org/?CMD=info&NAME=deh_cptr_actor < Actor codepointers
http://www.zdaemon.org/?CMD=info&NAME=deh_cptr_weapon < Weapon codepointers
http://www.zdaemon.org/?CMD=info&NAME=deh_thingflags < Thing flags

Also there is by now official dehsupp, which doomer will provide hopefully.

**Comment by GitExl** [comment]
Great, those lists are perfect! I've made some headway with ZDaemon support a few months ago, most things are done except for extended thing flag editing and some testing\polish. It may still be a while, but I'll get around to it sooner or later.

**Comment by GitExl** [comment]
I've finished working on the extended features. There's a beta release that includes a ZDaemon configuration that supports all the new features in ZDaemon thus far: https://github.com/GitExl/WhackEd4/releases/tag/v1.1.0-beta

Since a lot of internal things have changed, there will certainly be bugs and such, so keep that in mind when you test it out. Also please let me know if I got the ZDaemon features right!

**Comment by Aeyesx** [comment]
This is great!
Still some things changes cuz there is stormy development in B07 that brings some decorate codepointers into dehacked using extended Args, which is awesome, yet doomer is busy by studying while I am kind of busy because of new job these days, so ill put this on zdaemon forum at least and ill try to collect more info and stuff. thanks!

**Comment by Aeyesx** [comment]
An awesome piece of work!
The options are just tip-top containing everything so far I've seen.
I'll put this on ZDaemon forum later, I have to thanks you a lot this will be very very handy for those who are not familiar with manual zdaemon mapping.

By the way... 
About DehSUPP it does not contain any extended states/weapons(up to 31)/or things...
Why?
It makes sence to me since there is no official one. (Which is kinda shame, but there's one already ready maintained by doomer but it's nowhere announced so who knows how this will end).
Yet, when I imagine Slade...
Would it be possible to load dehsupp source or content or even better convert those?
(this is already real luxury)
whatever lets see

**Comment by GitExl** [comment]
Adding things from a DEHSUPP lump is something I'll probably do in a later version. Using an "official" DEHSUPP as the source for the data would be gooed, but perhaps I can create a script or tool that can convert any DEHSUPP lump into a useable WhackEd table. But that's for a next version.

**Comment by Aeyesx** [comment]
That would be very very cool,

Imagine how time passes many people coded various monsters/weapons/objects based on different dehsupps's,
This is its negative side, since you cannot use those anymore or mix them up. There's TNSTeam library that suffers a lot from it, but re-writing it would take month.

**Comment by Aeyesx** [comment]
ZD Beta07 went out, i guess itll slow down by now, how nice to have things in one place.

**Comment by MaiklRussia** [comment]
Hello everyone! :)
GitExl, Is 1.1.0 version of winhacked not support DEHSUPP yet?

**Comment by GitExl** [comment]
Thanks for the report MaiklRussie. To fix that, replace whacked4\cfg\tables_zdaemon.json with the one from this ZIP file to fix that: [tables_zdaemon.zip](https://github.com/GitExl/WhackEd4/files/67445/tables_zdaemon.zip)

There's a similar bug in the MBF configuration, this JSON file fixes that one: [tables_mbf.zip](https://github.com/GitExl/WhackEd4/files/67447/tables_mbf.zip)

**Comment by GitExl** [comment]
Not yet, no. I've reserved that for 1.2, but I don't know when I'll get
around to working on that.

**Comment by MaiklRussia** [comment]
BTW, is it possible add Heretic to winhacked?

**Comment by Aeyesx** [comment]
Dude, doomer's official ZD dehsupp already supports all states, however Whacked4 cannot load any Dehsupp atm. but by default it might be able to so it can convert old ones into new one. That's why I was tellin ya not to use JCD's one.

**Comment by MaiklRussia** [comment]
I mean vanilla Heretic, not Heretic actors in Doom.

**Comment by Aeyesx** [comment]
Heretic or hexen is not anyhow related to Whacked4, WhackEd4 is just a DeHackEd editor...All what is related here is just a game-type that u can specify for actors.
